src/XrdNet/XrdNetUtils.cc

```diff
@@ -938,3 +940,72 @@ bool XrdNetUtils::Singleton(const char  *hSpec, const char **eText)
 //
    return isSingle;
 }
+
+bool XrdNetUtils::ConnectWithTimeout(int sockfd, const struct sockaddr* clientAddr,size_t clientAddrLen, uint32_t timeout_sec, std::stringstream &errMsg) {
+
+  if (!SetSockBlocking(sockfd, false, errMsg)) {  // Set to non-blocking
+    close(sockfd);
+    return false;
+  }
+
+  int result = connect(sockfd, clientAddr, clientAddrLen);
+  if (result == 0) {
+    //We've managed to connect immediately
+    SetSockBlocking(sockfd, true, errMsg);  // Set the socket back to blocking
+    return true;
+  } else if (errno != EINPROGRESS) {
+    errMsg << "Connection failed: " << strerror(errno);
+    close(sockfd);
+    return false;
+  }
+
+  struct pollfd fds;
+  fds.fd = sockfd;
+  fds.events = POLLOUT;  // Wait for the socket to be ready to write
+
+  result = poll(&fds, 1, timeout_sec * 1000); //Timeout in milliseconds
+
+  if (result < 0) {
+    errMsg << "Poll error: " << strerror(errno);
+    ::close(sockfd);
+    return false;
+  } else if (result == 0) {
+    errMsg << "Connection timed out";
+    ::close(sockfd);
+    return false;
+  } else {
+    if (fds.revents & POLLOUT) {
+      int so_error;
+      socklen_t len = sizeof(so_error);
+      getsockopt(sockfd, SOL_SOCKET, SO_ERROR, &so_error, &len);
+      if (so_error == 0) {
+        // We managed to connect
+        SetSockBlocking(sockfd, true, errMsg);  // Set back to blocking
+        return true;
+      } else {
+        errMsg << "Connection failed after poll: " << strerror(so_error);
+        ::close(sockfd);
+        return false;
+      }
+    }
+  }
+  // We should not arrive here normally
+  return false;
+}
+
+bool XrdNetUtils::SetSockBlocking(int sockfd, bool blocking, std::stringstream & errMsg) {
+  int flags = fcntl(sockfd, F_GETFL, 0);
+  if (flags == -1) {
+    errMsg << "Failed to get socket flags " << strerror(errno);
+    return false;
+  }
+
+  flags = blocking ? (flags & ~O_NONBLOCK) : (flags | O_NONBLOCK);
+
+  if (fcntl(sockfd, F_SETFL, flags) == -1) {
+    errMsg << "Failed to set socket blocking/non-blocking";
```

Comment:
It's probably a good idea to add `<< strerror(errno)` on the set side as well, no?

Reply:
Done with 773d2469aa4a34e05d6d2ba61b656a8b35303db1